src/main/java/jenkins/scm/impl/avatars/UrlAvatarCacheSource.java

```diff
@@ -0,0 +1,137 @@
+            if (!canFetch()) {
+                LOGGER.log(Level.FINE, "Unable to fetch remote avatar: {0}", url);
+                return AvatarImage.EMPTY;
+            }
+            HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
+            try {
+                connection.setConnectTimeout(10000);
+                connection.setReadTimeout(30000);
+                if (connection.getResponseCode() >= 400) {
+                    LOGGER.log(Level.FINE, "Got invalid content response {1} for remote avatar image from {0}",
+                            new String[] { url, String.valueOf(connection.getResponseCode()) });
+                    return AvatarImage.EMPTY;
+                }
+                String contentType = connection.getContentType();
+                if (contentType == null || !contentType.startsWith("image/")) {
+                    LOGGER.log(Level.FINE, "Got invalid content type {1} for remote avatar image from {0}",
+                            new String[] { url, contentType });
+                    return AvatarImage.EMPTY;
+                }
+                int length = connection.getContentLength();
+                // buffered stream should be no more than 16k if we know the length
+                // if we don't know the length then 8k is what we will use
+                length = length > 0 ? Math.min(16384, length) : 8192;
+                InputStream is = null;
+                try {
+                    is = connection.getInputStream();
+                    BufferedInputStream bis = new BufferedInputStream(is, length);
+                    BufferedImage image = ImageIO.read(bis);
+                    if (image == null) {
+                        LOGGER.log(Level.FINE, "Got no remote avatar image from {0}", url);
+                        return AvatarImage.EMPTY;
+                    }
+                    return new AvatarImage(image, connection.getLastModified());
+                } catch (Exception e) {
+                    LOGGER.log(Level.INFO, "Failed to read from stream:" + e.getMessage(), e);
+                } finally {
+                    IOUtils.closeQuietly(is);
+                }
+            } catch (Exception e) {
+                LOGGER.log(Level.INFO, "Failed to connect to remote:" + e.getMessage(), e);
+            } finally {
+                connection.disconnect();
+            }
+        } catch (IOException e) {
+            LOGGER.log(Level.INFO, e.getMessage(), e);
+            return AvatarImage.EMPTY;
+        } finally {
+            long end = System.nanoTime();
+            long duration = TimeUnit.NANOSECONDS.toMillis(end - start);
+            LOGGER.log(duration > 250 ? Level.INFO : Level.FINE, "Avatar lookup of {0} took {1}ms",
+                    new Object[] { url, duration });
+        }
+        return AvatarImage.EMPTY;
+    }
+
+    /**
+     * Generate hash key used for caching
+     */
+    @Override
+    public String hashKey() {
```

Comment:
```suggestion
    public String getId() {
```

Sorry, I thought I'd said this in the first review.  This is not a hash, it is an ID string.  If we get a new items with the same hash it might just be a hash collision. If we get a new item with the same id, it should always replace the old item, right?

Reply:
@mlasevich 
https://github.com/mlasevich/scm-api-plugin/pull/1

Reply:
fair enough. :-)